internal/config/variables/variables.go

```diff
@@ -0,0 +1,531 @@
+// provided file, using the provided source to set the pb.Variable.Source value.
+func parseFileValues(filename string, source string) ([]*pb.Variable, hcl.Diagnostics) {
+	var pbv []*pb.Variable
+	f, diags := readFileValues(filename)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+
+	attrs, moreDiags := f.Body.JustAttributes()
+	diags = append(diags, moreDiags...)
+	// We grab all variables here; we'll later check set variables against the
+	// known variables defined in the waypoint.hcl on the runner when we
+	// consolidate values from local + server
+	for name, attr := range attrs {
+		val, moreDiags := attr.Expr.Value(nil)
+		diags = append(diags, moreDiags...)
+
+		v := &pb.Variable{
+			Name: name,
+		}
+		// Set type
+		switch val.Type() {
+		case cty.String:
+			v.Value = &pb.Variable_Str{Str: val.AsString()}
+		case cty.Bool:
+			v.Value = &pb.Variable_Bool{Bool: val.True()}
+		case cty.Number:
+			var num int64
+			err := gocty.FromCtyValue(val, &num)
+			if err != nil {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid number",
+					Detail:   err.Error(),
+					Subject:  &attr.Range,
+				})
+				return nil, diags
+			}
+			v.Value = &pb.Variable_Num{Num: num}
+		}
+
+		// Set source
+		switch source {
+		case sourceFile:
+			v.Source = &pb.Variable_File_{}
+		case sourceVCS:
+			v.Source = &pb.Variable_Vcs{}
+		}
+		pbv = append(pbv, v)
+	}
+
+	return pbv, diags
+}
+
+// readFileValues is a helper function that loads a file, parses if it is
+// hcl or json, and checks for any errant variable definition blocks. It returns
+// the files contents for further evaluation.
+func readFileValues(filename string) (*hcl.File, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+
+	// load the file
+	src, err := ioutil.ReadFile(filename)
+	if err != nil {
+		var errStr string
+		if os.IsNotExist(err) {
+			errStr = fmt.Sprintf("Given variables file %s does not exist.", filename)
+		} else {
+			errStr = fmt.Sprintf("Error while reading %s: %s.", filename, err)
+		}
+		return nil, append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Failed to read variable values from file",
+			Detail:   errStr,
+		})
+	}
+
+	// parse the file, whether it's hcl or json
+	var f *hcl.File
+	if strings.HasSuffix(filename, ".json") {
+		var hclDiags hcl.Diagnostics
+		f, hclDiags = hcljson.Parse(src, filename)
+		diags = append(diags, hclDiags...)
+		if f == nil || f.Body == nil {
+			return nil, diags
+		}
+	} else {
+		var hclDiags hcl.Diagnostics
+		f, hclDiags = hclsyntax.ParseConfig(src, filename, hcl.Pos{Line: 1, Column: 1})
+		diags = append(diags, hclDiags...)
+		if f == nil || f.Body == nil {
+			return nil, diags
+		}
+	}
+
+	// Before we do our real decode, we'll probe to see if there are any
+	// blocks of type "variable" in this body, since it's a common mistake
+	// for new users to put variable declarations in wpvars rather than
+	// variable value definitions.
```

Comment:
This is great diagnostics work!

Reply:
I will take the credit for stealing this idea 😄